hazelcast/src/main/java/com/hazelcast/nio/tcp/FastSelectionKeys.java

```
@@ -0,0 +1,210 @@
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.util.AbstractSet;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static java.lang.Class.forName;
+import static java.lang.System.arraycopy;
+
+/**
+ * The FastSelectionKeys optimized the Selector so less litter is being created. The Selector uses a Set, but this creates
+ * an object for every add of a selection key. With this FastSelectionKeys an array is being used since every key is going
+ * to be inserted only once.
+ *
+ * This trick comes from Netty.
+ */
+public final class FastSelectionKeys extends AbstractSet<SelectionKey> {
+    static final int INITIAL_CAPACITY = 32;
+    static final String SELECTOR_IMPL = "sun.nio.ch.SelectorImpl";
+
+    // Using this property the FastSelectionKeys optimization can be disabled. It is enabled by default.
+    private static final boolean DISABLED = Boolean.getBoolean("hazelcast.io.fastselectionkeys.disabled");
+
+
+    // the active SelectionKeys is the one where is being added to.
+    SelectionKeys activeKeys = new SelectionKeys();
+    // the passive SelectionKeys is one that is being read using the iterator.
+    SelectionKeys passiveKeys = new SelectionKeys();
+    private final IteratorImpl iterator = new IteratorImpl();
+
+    FastSelectionKeys() {
+    }
+
+    @Override
+    public boolean add(SelectionKey o) {
+        return activeKeys.add(o);
+    }
+
+    @Override
+    public int size() {
+        return activeKeys.size;
+    }
+
+    @Override
+    public Iterator<SelectionKey> iterator() {
+        iterator.init(flip());
+        return iterator;
+    }
+
+    private SelectionKey[] flip() {
+        SelectionKeys tmp = activeKeys;
+        activeKeys = passiveKeys;
+        passiveKeys = tmp;
+
+        activeKeys.size = 0;
+        return passiveKeys.keys;
+    }
+
+    @Override
+    public boolean remove(Object o) {
```

Is this not-implemented because it's not called? In that case, I think it's better to throw unsupported exception instead of no-op.